### DISCUSSION TITLE
Adding state to State

### DISCUSSION THREAD

**Discussion by Sagebati** [mitsuhiko/minijinja#530] [Q&A]
Hi, 

Big user of minijinja, I'm in the need of using async functions in the `add_function`. I saw different examples, with impl Object. I find this solution difficult and I think most of it could be resolved if we added custom state a la axum. `Environement::<MyCustomState>::with_state(env,state)`

```
pub fn custom_filter(s: &State<MyCustomState>, _v: Value) -> Result<Value, Error> {
     s.state // We can access the my custom state        
}
```

Is this something that could be integrated or it's a no go ?

**Comment by mitsuhiko** [comment]
The big question with this is: what would be the lifetime of that? Today the state's lifetime is a bit complicated as for instance extending, including and macro invocations will crate their own state.

**Reply by Sagebati** [reply]
If by "that" you mean the type `MyCustomState`,  the idea is that the custom state would be added to `Environment`, with for `add_custom_state()` so it lives as least as long as the `Environment`. For me the problem is that we have to add a type generic to many structs.
